Enumerable-Project-Ruby: Enumerable feature

brugobi opened the pull request:
## Enumerables Methods

- The methods in this code are enumerables to work with arrays.

### it contains copies of all the most used enumerables:  
- each
- each_with_index
- my_select
- my_all?
- my_any?
- my_none?
- my_count
- my_map
- my_inject

### thank you for your review

brugobi committed:
Set-up Stickler

brugobi committed:
add an empty file for ruby code

brugobi committed:
add my each

asiimwedismas committed:
Refactored the my_each method

brugobi committed:
add my_each_with_index

asiimwedismas committed:
Created the my_select method

asiimwedismas committed:
Refactored the my_all method

brugobi committed:
add my all

brugobi committed:
add my_any

asiimwedismas committed:
Created the my_none method

asiimwedismas committed:
Minor change

brugobi committed:
add my_count

asiimwedismas committed:
Added the my_map method

asiimwedismas committed:
Updated readme file

asiimwedismas committed:
Updated readme file

brugobi committed:
refactory my_inject function

asiimwedismas committed:
Removed my_map that takes either the proc or block

asiimwedismas committed:
Refactored my_count to work with and without an argument and a block

brugobi committed:
creating a method called #multiply_els to test my_inject method

asiimwedismas committed:
Merge branch 'enumerable-feature' of https://github.com/brugobi/Enumerables-Ruby into enumerable-feature

asiimwedismas committed:
Modified my_map method to take either a block or a proc

brugobi committed:
Modify my_map method to take either a proc or a block

asiimwedismas committed:
Added gemfile and made some minor change to test.rb

asiimwedismas committed:
Fixed rubocop linters

asiimwedismas committed:
Added a few tests for my_inject method

asiimwedismas committed:
Updated the readme file

asiimwedismas committed:
Minor change to force stickler to refresh

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
```
### **my_each** 

- [x] When a block is given after calling the given block once for each element in self, returns the array itself.

asiimwedismas replied:
fixed

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
```
### **my_each_with_index** 

- [x] When a block is given after calling the given block once for each element in self, returns the array itself

asiimwedismas replied:
fixed

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_all
+    return true unless block_given?
```
### **my_all**

- [x] It should return true if the block never returns false or nil

- [x] It should not mutate the original array

- [x]  When no block or argument is given it should return true when none of the collection members are false or nil

- [x] When a class is passed as an argument returns true if all of the collection is a member of such class

- [x] When a Regex is passed as an argument returns true if all of the collection matches the Regex

- [x] When a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

asiimwedismas replied:
all ruby doc tests passed, no coded changed
![test all](https://user-images.githubusercontent.com/51087315/83947958-eb676e00-a822-11ea-8dee-0bb03f47f91f.png)

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_any
+    return true unless block_given?
```
### **my_any**

- [x] It should return true if the block ever returns a value other than false or nil

- [x] It should not mutate the original array

- [x] When no block or argument is given it should return true if at least one of the collection is not false or nil

- [x] When a class is passed as an argument it should return true if at least one of the collection is a member of such class

- [x] When a Regex is passed as an argument it should return false if none of the collection matches the Regex

- [x] When a pattern other than Regex or a Class is given it should return false if none of the collection matches the pattern

asiimwedismas replied:
all ruby doc tests passed 
![passed2](https://user-images.githubusercontent.com/51087315/83947883-62e8cd80-a822-11ea-9270-c459d431fb07.png)

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_none
+    return true unless block_given?
```
### **my_none**

- [x] It should return true if the block never returns true for all elements.

- [x] It should not mutate the original array

- [x] when no block or argument is given it should return true only if none of the collection members is true

- [x] when a class is passed as an argument it should return true if none of the collection is a member of such class

- [x] when a Regex is passed as an argument it should return true only if none of the collection matches the Regex

asiimwedismas replied:
fixed the logic, all ruby doc test passed 
![passed2](https://user-images.githubusercontent.com/51087315/83947787-e229d180-a821-11ea-9127-e45c8e242232.png)

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_count(arg = nil)
+    counter = 0
```
### **my_count**

- [x] It should return the number of items in enum through enumeration

asiimwedismas replied:
issue fixed

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+  end
+
+  def my_map(a_proc = nil)
+    return to_enum unless block_given?
```
### **my_map**

- [x] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum

asiimwedismas replied:
Ruby won't let us supply both a proc and block at the same time, so I am not sure how we are supposed to do this if it cant run in the first place when both are supplied
![error](https://user-images.githubusercontent.com/51087315/83785011-584efc80-a699-11ea-8809-94fc69effe75.jpg)

serdg0 commented on enumerable.rb:
```diff
@@ -0,0 +1,89 @@
+
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+  def my_inject(*args)
+    args_size = args.length
```
### **my_inject**

- [x] combines all elements of enum by applying a binary operation, specified by a block

- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

brugobi replied:
the test passes all the test on the ruby documentation

serdg0 reviewed (changes requested):
### Hello! @brugobi @asiimwedismas :peace_symbol: 

Great job! You are close to the finish line! :racing_car: :racing_car: 

This project requires to consider some edge cases, which I point below!

After adressing those issues, submit for another CRR.

**Happy Coding!** :computer: :nerd_face:
**_Sergio Diaz_**

asiimwedismas committed:
fixed my_each
When a block is given after calling the given block once for each element in self, returns the array itself.

asiimwedismas committed:
Fixed my_each_index

When a block is given after calling the given block once for each element in self, returns the array itself

asiimwedismas committed:
Fixed my_Count now returns size of the array if no block is given

brugobi committed:
updated my_each method to return the original array

brugobi committed:
updated my_each_with_index  method to return the original array

brugobi committed:
updated my_all method

brugobi committed:
updated my_any and my_all

brugobi committed:
updated my_count

brugobi commented:
all the issues were fixed

## thank you for the review

brugobi committed:
updated the README

asiimwedismas committed:
Merge branch 'enumerable-feature' of https://github.com/brugobi/Enumerables-Ruby into enumerable-feature

asiimwedismas committed:
updated my_each to accept ranges as well

Dandush03 commented on enumerable.rb:
```diff
@@ -0,0 +1,124 @@
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    my_array = is_a?(Range) ? to_a : self
+
+    my_array.size.times { |n| yield (my_array[n]) }
+    self
+  end
```
### #my_each 
- [x] when a block is given and self is a hash behaves like #each calls the given block once for each element in self

asiimwedismas replied:
Already works!!!!!!!!!

> # Status: 🔴 Required Changes 🔴
> ## Hey! @brugobi 👋
> In order to get your code to work, we need to make sure that #my_each works as it should. otherwise, all the rest of your function will have the wrong behavior.
> 
> right now when testing your code I get an error when a block is given and self is a hash; instead of receiving the given block once for each element in self, I got an empty string.
> 
> so in order to make it work properly, we need to use 2 different conditions, the first one will check if block is given and the second one will check the method used, otherwise as you already know should return `self` which returns an enum hash. right now you are only checking for the Range method while missing the other
> 
> If you need some extra help you can write to me through slack.
> 
> ### now lest dive in your code.
> ## Please make sure to fix all issues pointed below before submitting a new review.

Please share your test cases

Dandush03 commented on enumerable.rb:
```diff
@@ -0,0 +1,124 @@
+    self
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    my_array = is_a?(Range) ? to_a : self
+
+    my_array.size.times { |n| yield my_array[n], n }
+    self
+  end
```
### #my_each_
* [x] with_index when a block is given and self is a hash behaves like #each_with_index calls the given block once for each element in self

asiimwedismas replied:
ALREADY WORKS !!!!!!! THIS PASSED THE EXAMPLE TEST
![error2](https://user-images.githubusercontent.com/51087315/83947096-937a3880-a81d-11ea-9f38-d4d24979c2f8.jpg)
 FROM THE RUBY DOCUMENTATION, WHAT OTHER TEST IS IT SUPPOSED TO PASS???

Dandush03 commented on enumerable.rb:
```diff
@@ -0,0 +1,124 @@
+  def my_all(arg = nil)
+    result = true
+
+    my_each do |i|
```
### #my_all? 
* [x] All Boolean function should end with a question mark for best practices

asiimwedismas replied:
Fixed

Dandush03 commented on enumerable.rb:
```diff
@@ -0,0 +1,124 @@
+    result
+  end
+
+  def my_none(arg = nil)
+    result = true
+    my_each do |i|
+      result = if arg
+                 arg.is_a?(Class) ? i.is_a?(arg) : i.to_s.match?(arg.to_s)
+               elsif block_given?
+                 !yield i
+               else
+                 !i == false || i.nil? ? false : true
+               end
+      return true if result
+    end
+    result
+  end
```
### #my_none? 
* [x] All Boolean function should end with a question mark for best practices
* [x] Returns false if the block ever returns true for all elements in a range
* [x] When no block or argument is given returns false only if one of the collection members is true
* [x] When a Regex is passed as an argument returns true if none of the collection matches the Regex
* [x] When a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
* [x] When a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

asiimwedismas replied:
fixed the logic
![passed2](https://user-images.githubusercontent.com/51087315/83947999-2c5f8280-a823-11ea-8f5d-0b0d5df3da1d.png)

asiimwedismas replied:
all rub doc tests passed

Dandush03 commented on enumerable.rb:
```diff
@@ -0,0 +1,124 @@
+    counter
+  end
+
+  def my_map(a_proc = nil)
+    return to_enum unless block_given?
+
+    map_array = []
+
+    if a_proc.nil?
+      my_each { |n| map_array << yield(n) }
+    else
+      my_each { |n| map_array << proc.call(n) }
+    end
+
+    map_array
+  end
```
### #my_map 
* [x] executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

asiimwedismas replied:
this is impossible, ruby won't let you pass both a block
![error](https://user-images.githubusercontent.com/51087315/83946658-eb637000-a81a-11ea-9edc-a0c5bebbaf6c.jpg)
 and a proc

Dandush03 commented on test.rb:
```diff
@@ -0,0 +1,24 @@
+load 'enumerable.rb'
+
+def multiply_els(arr)
+  arr.my_inject { |sum, n| sum * n }
+end
```
### #multiply_els
* [x] Accepts an array as an argument and multiplies all the elements of the array together using #my_inject

asiimwedismas replied:
please check the test.rb file for this test case

Dandush03 reviewed (commented):
# Status:  🔴 Required Changes 🔴 

## Hey! @brugobi 👋 

In order to get your code to work, we need to make sure that #my_each works as it should. otherwise, all the rest of your function will have the wrong behavior.

right now when testing your code I get an error when a block is given and self is a hash; instead of receiving the given block once for each element in self, I got an empty string.

so in order to make it work properly, we need to use 2 different conditions, the first one will check if block is given and the second one will check the method used, otherwise as you already know should return `self` which returns an enum hash. right now you are only checking for the Range method while missing the other

If you need some extra help you can write to me through slack.

------

### now lest dive in your code.

-----
## Please make sure to fix all issues pointed below before submitting a new review.

asiimwedismas committed:
Added ? to all boolean functions

asiimwedismas committed:
added test cases for tse

asiimwedismas committed:
fixed logic of my_none to pass ruby documentation test

asiimwedismas committed:
All rasied issues have been fixed

asiimwedismas commented:
- boolean functions now have ? at the end
- the logic for my_none has been fixed to pass all the ruby documentation test

MarioDena reviewed (changes requested):
Hello, it seems that you are misunderstanding a few requirements. the project can't be approved because the code is not behaving as it should be. Remember all TSE did this project and had to pass all requirements.

Ill list the failing scenarios found, but it's up to you to find the answers, if you answer with "It's impossible" you are not learning, imagine this is asked in a working environment, "It's impossible" is not a valid answer when changes are required on your code. I'm telling you this because I understand your frustration and I started to think as if my boss is asking for changes, So with that new mindset I used my time and resources to find the answer and make my code pass all test. 

So here is a layout of the failed scenarios. (All where previously asked changes)

- [x] #my_each should return an Enumerator if no block is given. This is not happening.
- [x] #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self. This case is returning an empty string. 
- [x] #my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self.

- [ ] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum. I wanted to expand on this since your answer was that "it's imposible" Ruby has error handling of course but doesnt that tell's you that then ruby somehow has a way to take both arguments and since your code doesn't handle that scenario the default error is displayed? That's as much as a clue I can give you, but not only is this possible several students have made several different code examples, so it's even possible in multiple ways.

- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject This gives a no method error.

I hope you can take my words as a learning experience, and can find the right mindset to overcome this challenge.

Cheers! 🎖️

asiimwedismas committed:
my_each now returns an enumerable when no block is given

asiimwedismas commented:
@MarioDena, @Dandush03, @serdg0 , I still cant' figure out what you mean by
 # my_each when a block is given when self is a hash behaves like #each calls the given block 
 once for each element in self. This case is returning an empty string. 

 ... Especially since all our code is pretty much similar or even better than ALL the other student's code including yours, please show us how to accomplish this, thank you

asiimwedismas commented:
** my_each_with_index when a block is given when self is a hash behaves like #each_with_index calls the given block once for each element in self. **

@MarioDena  once again, our code is same as your (logically) so please do help out.  Actually yours is incomplete since it is supposed to return the original array

asiimwedismas commented:
#multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject This gives a no method error.

I stated earlier on that this method was in the test.rb file but I have moved it to the enumerable.rb file so you can find it.

asiimwedismas commented:
#multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject This gives a no method error.

I stated earlier on that this method was in the test.rb file but I have moved it to the enumerable.rb file so you can find it.

asiimwedismas committed:
Moved multiply_els method from test.rb file to enumerable.rb file

asiimwedismas commented:
@MarioDena @serdg0 @Dandush03 , i havelooked at all your enumerables project to see how at least one of you completed your code so my_map accepts both a proc and a block and there literally is no difference between our code and yours.

I decided to copy @MarioDena into our project and run it with both a proc and a block and ruby still won't accept it... My partner asked her stand up members about this same issue and they said this issue come up and when they pointed out that ruby wouldn't accept both proc and block, their code was approved...

I love coding and neat tricks... I humbly request any of you three to show me this ruby hack so I can pass both a proc and block. 

Thank you

![mario](https://user-images.githubusercontent.com/51087315/83961981-6c654a80-a8a1-11ea-9652-da854ed4663f.jpg)

on the left is @MarioDena code from his master branch, I don't see any error handling in the method. And on the right is our code for the same method. The terminal shows the output of calling @MarioDena  #my_map method with both a proc and block

asiimwedismas committed:
Moved multiply_els to enumerable.rb

asiimwedismas committed:
Added TSE master branch code to test his my_map method

asiimwedismas committed:
Commented out my_map call in marioDena-enumberbles.rb

asiimwedismas committed:
Deleted marioDena enumberables because rubocop wasn't happy with it

MarioDena reviewed (changes requested):
Hello again!

I can't approve your project until you have fixed the non passing test. If you are unable to find a solution by yourself you can contact me on slack and I can help you figure it out. (But don't request a code review since we TSE will have to request chances and wont approve it until the project has everything needed, contact me I can help you and the submit, or you can contact student success or anyone else to help you of course you are not forced to contact me only)

I would also like to say that all my comments are meant to help you, as I read maybe you were angry since you think. " Especially since all our code is pretty much similar or even better than ALL the other student's code including yours, please show us how to accomplish this, thank you" 

Remember that "better" code is subjective. Besides you are not competing against anyone, you just need to learn for you and your future job. Of course my code is outdated since microverse is improving the curriculum every day, But I have made the changes locally to pass the new test, and I can guarantee you it's possible. 

If you choose to contact me to help you figure out stuff Ill be glad to help, but again only submit a code review when your project is ready since asking for a code review is not the correct channel to look for help. (There is student success, your mentor, the TSE associated with the review that you can contact on slack or on the dashboard).

I understand your frustration, but just keep going you will made it since you are a talented programmer.

Cheers! 😄

asiimwedismas committed:
Updated my_map to execute only proc when both proc block are supplied

asiimwedismas commented:
> Hello again!
> 
> I can't approve your project until you have fixed the non passing test. If you are unable to find a solution by yourself you can contact me on slack and I can help you figure it out. (But don't request a code review since we TSE will have to request chances and wont approve it until the project has everything needed, contact me I can help you and the submit, or you can contact student success or anyone else to help you of course you are not forced to contact me only)
> 
> I would also like to say that all my comments are meant to help you, as I read maybe you were angry since you think. " Especially since all our code is pretty much similar or even better than ALL the other student's code including yours, please show us how to accomplish this, thank you"
> 
> Remember that "better" code is subjective. Besides you are not competing against anyone, you just need to learn for you and your future job. Of course my code is outdated since microverse is improving the curriculum every day, But I have made the changes locally to pass the new test, and I can guarantee you it's possible.
> 
> If you choose to contact me to help you figure out stuff Ill be glad to help, but again only submit a code review when your project is ready since asking for a code review is not the correct channel to look for help. (There is student success, your mentor, the TSE associated with the review that you can contact on slack or on the dashboard).
> 
> I understand your frustration, but just keep going you will made it since you are a talented programmer.
> 
> Cheers! 😄

updated, problem fixed

lekegitrepo commented:
#### Hi @brugobi and @asiimwedismas :clap: :clap:
#### Good job :+1: This project is approved.

### _Reviewed by Temitayo Adeleke._

lekegitrepo reviewed (approved):
#### Status: Approved.

asiimwedismas committed:
Added ruby doc tests and performed some code refactoring
